shared/src/main/java/uk/gov/di/authentication/shared/services/AuthSessionService.java

```diff
@@ -109,6 +110,30 @@ public Optional<AuthSessionItem> getSession(String sessionId) {
         return validAuthSession;
     }
 
+    public Optional<AuthSessionItem> getSessionFromRequestHeaders(Map<String, String> headers) {
+        Optional<String> sessionId =
+                getOptionalHeaderValueFromHeaders(
+                        headers,
+                        SESSION_ID_HEADER,
+                        configurationService.getHeadersCaseInsensitive());
+
+        if (sessionId.isEmpty()) {
+            LOG.warn("Value not found for Session-Id header");
+        }
+
+        return sessionId
+                .flatMap(InputSanitiser::sanitiseBase64)
+                .flatMap(
+                        id -> {
+                            try {
+                                return getSession(id);
+                            } catch (Exception e) {
+                                LOG.error("Failed to retrieve session for id: {}", id);
+                                throw e;
```

Comment:
Maybe define and throw a dedicated exception (eg. `AuthSessionServiceException`) instead?

Reply:
Nice, didn't know this exception existed, have implemented this!